crates/red_knot/src/lint.rs

```diff
@@ -57,6 +62,108 @@ pub(crate) fn lint_lines(source: &str, diagnostics: &mut Vec<String>) {
     }
 }
 
+#[tracing::instrument(level = "debug", skip(db))]
+pub(crate) fn lint_semantic<Db>(db: &Db, file_id: FileId) -> Diagnostics
+where
+    Db: SemanticDb + HasJar<SemanticJar>,
+{
+    let storage = &db.jar().lint_semantic;
+
+    storage.get(&file_id, |file_id| {
+        let source = db.source(*file_id);
+        let parsed = db.parse(*file_id);
+        let symbols = db.symbol_table(*file_id);
+
+        let context = SemanticLintContext {
+            file_id: *file_id,
+            source,
+            parsed,
+            symbols,
```

Comment:
What's the advantage of putting these directly in the context, rather than just giving the lint rule the db and letting it ask for what it needs?

Reply:
A large portion of lint rules need access to these struct. Resolving them once reduces the number of hash table lookups and arc clones that are needed.

I also don't want to give lint rules access to `db`. I think we instead want to expose some methods on context (e.g. I don't know if I want lint rules to read the source text of other files).